aggregator/elem.go

```diff
@@ -192,17 +194,27 @@ func (e *CounterElem) Consume(earlierThanNanos int64, fn aggMetricFn) bool {
 		if e.values[idx].timeNanos >= earlierThanNanos {
 			break
 		}
-		endAtNanos := e.values[idx].timeNanos + int64(e.policy.Resolution().Window)
-		e.processValue(endAtNanos, e.values[idx].counter, fn)
 		idx++
 	}
+	e.toConsume = e.toConsume[:0]
 	if idx > 0 {
 		// Shift remaining values to the left and shrink the values slice
+		e.toConsume = append(e.toConsume, e.values[:idx]...)
 		n := copy(e.values[0:], e.values[idx:])
 		e.values = e.values[:n]
 	}
 	canCollect := len(e.values) == 0 && e.tombstoned
 	e.Unlock()
+
+	// Process the counters.
+	for i := range e.toConsume {
+		endAtNanos := e.toConsume[i].timeNanos + int64(e.policy.Resolution().Window)
+		e.toConsume[i].counter.Lock()
+		e.processValue(endAtNanos, e.toConsume[i].counter.Counter, fn)
```

Comment:
Unfortunately we'll need it when we support custom aggregation functions (min/max/avg/etc)